ext/standard/ftp_fopen_wrapper.c

```diff
@@ -720,18 +720,17 @@ php_stream * php_stream_ftp_opendir(php_stream_wrapper *wrapper, const char *pat
 	if (hoststart == NULL) {
 		hoststart = resource->host;
 	}
-	datastream = php_stream_sock_open_host(hoststart, portno, SOCK_STREAM, 0, 0);
-	if (datastream == NULL) {
-		goto opendir_errexit;
-	}
 
 	result = GET_FTP_RESULT(stream);
 	if (result != 150 && result != 125) {
 		/* Could not retrieve or send the file
 		 * this data will only be sent to us after connection on the data port was initiated.
 		 */
```

Comment:
I looked around a bit today... The RFC is, as usual, underspecified. It looks like DJB has some more practical coverage on the topic.

For NLST/LIST it [says](https://cr.yp.to/ftp/list.html) (emphasis mine):

> A LIST or NLST request asks the server to send the contents of a directory over the data connection **already established by the client**.

Also, the [page on PASV](https://cr.yp.to/ftp/retr.html) says:

>  The server normally accepts PASV with code 227. Its response is a single line showing the IP address of the server and the TCP port number where the server is accepting connections.
>
> Normally the client will connect to this TCP port, from the same IP address that the client is using for the FTP connection, and then send a RETR request. However, the client may send some other requests first, such as REST. The server must continue to read and respond to requests while it accepts connections. Most operating systems handle this automatically.

Both of these seem to imply that the data connection is supposed to be opened before the RETR/NLST/... request is sent. The [example](https://cr.yp.to/ftp/browsing.html) does this as well. I also checked the ext/ftp code and unless I missed something, it also establishes the data connection before sending the request. See in particular this part of ftp_genlist: https://github.com/php/php-src/blob/master/ext/ftp/ftp.c#L1805

Maybe the fix is to open the data connection earlier (before sending the request) instead of later? Or am I just misunderstanding something?